# JSON examples and exercise
****
+ get familiar with packages for dealing with JSON
+ study examples with JSON strings and files 
+ work on exercise to be completed and submitted 
****
+ reference: http://pandas.pydata.org/pandas-docs/stable/io.html#io-json-reader
****

In [1]:
import pandas as pd

## imports for Python, Pandas

In [2]:
import json
from pandas.io.json import json_normalize

## JSON example, with string

+ demonstrates creation of normalized dataframes (tables) from nested json string
+ source: http://pandas.pydata.org/pandas-docs/stable/io.html#normalization

In [3]:
# define json string
data = [{'state': 'Florida', 
         'shortname': 'FL',
         'info': {'governor': 'Rick Scott'},
         'counties': [{'name': 'Dade', 'population': 12345},
                      {'name': 'Broward', 'population': 40000},
                      {'name': 'Palm Beach', 'population': 60000}]},
        {'state': 'Ohio',
         'shortname': 'OH',
         'info': {'governor': 'John Kasich'},
         'counties': [{'name': 'Summit', 'population': 1234},
                      {'name': 'Cuyahoga', 'population': 1337}]}]

In [4]:
# use normalization to create tables from nested element
json_normalize(data, 'counties')

name  population
0        Dade       12345
1     Broward       40000
2  Palm Beach       60000
3      Summit        1234
4    Cuyahoga        1337

In [5]:
# further populate tables created from nested element
json_normalize(data, 'counties', ['state', 'shortname', ['info', 'governor']])

name  population    state shortname info.governor
0        Dade       12345  Florida        FL    Rick Scott
1     Broward       40000  Florida        FL    Rick Scott
2  Palm Beach       60000  Florida        FL    Rick Scott
3      Summit        1234     Ohio        OH   John Kasich
4    Cuyahoga        1337     Ohio        OH   John Kasich

****
## JSON example, with file

+ demonstrates reading in a json file as a string and as a table
+ uses small sample file containing data about projects funded by the World Bank 

In [6]:
# load json as string
json.load((open('data/world_bank_projects_less.json')))

[{'_id': {'$oid': '52b213b38594d8a2be17c780'},
  'approvalfy': 1999,
  'board_approval_month': 'November',
  'boardapprovaldate': '2013-11-12T00:00:00Z',
  'borrower': 'FEDERAL DEMOCRATIC REPUBLIC OF ETHIOPIA',
  'closingdate': '2018-07-07T00:00:00Z',
  'country_namecode': 'Federal Democratic Republic of Ethiopia!$!ET',
  'countrycode': 'ET',
  'countryname': 'Federal Democratic Republic of Ethiopia',
  'countryshortname': 'Ethiopia',
  'docty': 'Project Information Document,Indigenous Peoples Plan,Project Information Document',
  'envassesmentcategorycode': 'C',
  'grantamt': 0,
  'ibrdcommamt': 0,
  'id': 'P129828',
  'idacommamt': 130000000,
  'impagency': 'MINISTRY OF EDUCATION',
  'lendinginstr': 'Investment Project Financing',
  'lendinginstrtype': 'IN',
  'lendprojectcost': 550000000,
  'majorsector_percent': [{'Name': 'Education', 'Percent': 46},
   {'Name': 'Education', 'Percent': 26},
   {'Name': 'Public Administration, Law, and Justice', 'Percent': 16},
   {'Name': 'Educatio

In [7]:
# load as Pandas dataframe
sample_json_df = pd.read_json('data/world_bank_projects_less.json')
sample_json_df

_id  approvalfy board_approval_month  \
0  {'$oid': '52b213b38594d8a2be17c780'}        1999             November   
1  {'$oid': '52b213b38594d8a2be17c781'}        2015             November   

      boardapprovaldate                                 borrower  \
0  2013-11-12T00:00:00Z  FEDERAL DEMOCRATIC REPUBLIC OF ETHIOPIA   
1  2013-11-04T00:00:00Z                    GOVERNMENT OF TUNISIA   

            closingdate                              country_namecode  \
0  2018-07-07T00:00:00Z  Federal Democratic Republic of Ethiopia!$!ET   
1                   NaN                      Republic of Tunisia!$!TN   

  countrycode                              countryname countryshortname  \
0          ET  Federal Democratic Republic of Ethiopia         Ethiopia   
1          TN                      Republic of Tunisia          Tunisia   

                         ...                           sectorcode source  \
0                        ...                          ET,BS,ES,EP   IBRD   
1                        ...                                BZ,BS   IBRD   

   status  supplementprojectflg  \
0  Active                     N   
1  Active                     N   

                                              theme1  \
0      {'Name': 'Education for all', 'Percent': 100}   
1  {'Name': 'Other economic management', 'Percent...   

                                      theme_namecode themecode   totalamt  \
0      [{'name': 'Education for all', 'code': '65'}]        65  130000000   
1  [{'name': 'Other economic management', 'code':...     54,24          0   

  totalcommamt                                                url  
0    130000000  http://www.worldbank.org/projects/P129828/ethi...  
1      4700000  http://www.worldbank.org/projects/P144674?lang=en  

[2 rows x 50 columns]

****
## JSON exercise

Using data in file 'data/world_bank_projects.json' and the techniques demonstrated above,
1. Find the 10 countries with most projects
2. Find the top 10 major project themes (using column 'mjtheme_namecode')
3. In 2. above you will notice that some entries have only the code and the name is missing. Create a dataframe with the missing names filled in.

# Exploring the JSON data 

In [10]:
# Open the file and read the file
with open('data/world_bank_projects.json', 'r') as json_file:
    wb_json_data = json.load(json_file)

In [16]:
# Type details
print(type(wb_json_data))

#Length of the file
print(len(wb_json_data))

<class 'list'>
500


In [12]:
# Type details for column at 0 index
print(type(wb_json_data[0]))

<class 'dict'>


In [14]:
# exploring the data
print(wb_json_data[0].keys())

dict_keys(['sector', 'supplementprojectflg', 'projectfinancialtype', 'prodline', 'mjtheme', 'idacommamt', 'impagency', 'project_name', 'mjthemecode', 'closingdate', 'totalcommamt', 'id', 'mjsector_namecode', 'docty', 'sector1', 'lendinginstr', 'countrycode', 'sector2', 'totalamt', 'mjtheme_namecode', 'boardapprovaldate', 'countryshortname', 'sector4', 'prodlinetext', 'productlinetype', 'regionname', 'status', 'country_namecode', 'envassesmentcategorycode', 'project_abstract', 'approvalfy', 'projectdocs', 'lendprojectcost', 'lendinginstrtype', 'theme1', 'grantamt', 'themecode', 'borrower', 'sectorcode', 'sector3', 'majorsector_percent', 'board_approval_month', 'theme_namecode', 'countryname', 'url', 'source', 'projectstatusdisplay', 'ibrdcommamt', 'sector_namecode', '_id'])


In [17]:
#Normalization to create a table and examine the values
print(json_normalize(wb_json_data).head(10))

                   _id.$oid approvalfy board_approval_month  \
0  52b213b38594d8a2be17c780       1999             November   
1  52b213b38594d8a2be17c781       2015             November   
2  52b213b38594d8a2be17c782       2014             November   
3  52b213b38594d8a2be17c783       2014              October   
4  52b213b38594d8a2be17c784       2014              October   
5  52b213b38594d8a2be17c785       2014              October   
6  52b213b38594d8a2be17c786       2014              October   
7  52b213b38594d8a2be17c787       2014              October   
8  52b213b38594d8a2be17c788       2014              October   
9  52b213b38594d8a2be17c789       2014              October   

      boardapprovaldate                                 borrower  \
0  2013-11-12T00:00:00Z  FEDERAL DEMOCRATIC REPUBLIC OF ETHIOPIA   
1  2013-11-04T00:00:00Z                    GOVERNMENT OF TUNISIA   
2  2013-11-01T00:00:00Z   MINISTRY OF FINANCE AND ECONOMIC DEVEL   
3  2013-10-31T00:00:00Z   MIN. OF 

# Find the 10 countries with most projects
## Create the Dataframe from JSON file

In [18]:
#Create the dataframe from the JSON file
df = pd.read_json('data/world_bank_projects.json')

In [19]:
#Shape of the dataframe
print(df.shape)

#Columns of the dataframe
print(df.columns)

#Information of the dataframe
print(df.info())

#Descriptions of the dataframe
print(df.describe())

(500, 50)
Index(['_id', 'approvalfy', 'board_approval_month', 'boardapprovaldate',
       'borrower', 'closingdate', 'country_namecode', 'countrycode',
       'countryname', 'countryshortname', 'docty', 'envassesmentcategorycode',
       'grantamt', 'ibrdcommamt', 'id', 'idacommamt', 'impagency',
       'lendinginstr', 'lendinginstrtype', 'lendprojectcost',
       'majorsector_percent', 'mjsector_namecode', 'mjtheme',
       'mjtheme_namecode', 'mjthemecode', 'prodline', 'prodlinetext',
       'productlinetype', 'project_abstract', 'project_name', 'projectdocs',
       'projectfinancialtype', 'projectstatusdisplay', 'regionname', 'sector',
       'sector1', 'sector2', 'sector3', 'sector4', 'sector_namecode',
       'sectorcode', 'source', 'status', 'supplementprojectflg', 'theme1',
       'theme_namecode', 'themecode', 'totalamt', 'totalcommamt', 'url'],
      dtype='object')
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 500 entries, 0 to 499
Data columns (total 50 columns):
_id   

In [25]:
# Print the name of Countries with most projects
print(df['countryname'].value_counts().head(10))

People's Republic of China         19
Republic of Indonesia              19
Socialist Republic of Vietnam      17
Republic of India                  16
Republic of Yemen                  13
Kingdom of Morocco                 12
People's Republic of Bangladesh    12
Nepal                              12
Africa                             11
Republic of Mozambique             11
Name: countryname, dtype: int64


Above are the top 10 list of Countries with most project, as the count is on the distint values.

# Find the top 10 major project themes (using column 'mjtheme_namecode')

## Exploring the data in 'mjtheme_namecode' attribute 

In [38]:
# Print the shape of the column - 'mjtheme_namecode'
print(df['mjtheme_namecode'].shape)

# Print the head 2 records of the column - 'mjtheme_namecode'
print(df['mjtheme_namecode'].head(2))

# Print the if any null in the column - 'mjtheme_namecode'
print(df['mjtheme_namecode'].isna().sum())

# Print the first record of the column - 'mjtheme_namecode'
print(df['mjtheme_namecode'][0])

(500,)
0    [{'code': '8', 'name': 'Human development'}, {...
1    [{'code': '1', 'name': 'Economic management'},...
Name: mjtheme_namecode, dtype: object
0
[{'code': '8', 'name': 'Human development'}, {'code': '11', 'name': ''}]


In [36]:
# Checking the Key Values pair in the columns - 
for key , value in df['mjtheme_namecode'].items():
    print(key, value)

0 [{'code': '8', 'name': 'Human development'}, {'code': '11', 'name': ''}]
1 [{'code': '1', 'name': 'Economic management'}, {'code': '6', 'name': 'Social protection and risk management'}]
2 [{'code': '5', 'name': 'Trade and integration'}, {'code': '2', 'name': 'Public sector governance'}, {'code': '11', 'name': 'Environment and natural resources management'}, {'code': '6', 'name': 'Social protection and risk management'}]
3 [{'code': '7', 'name': 'Social dev/gender/inclusion'}, {'code': '7', 'name': 'Social dev/gender/inclusion'}]
4 [{'code': '5', 'name': 'Trade and integration'}, {'code': '4', 'name': 'Financial and private sector development'}]
5 [{'code': '6', 'name': 'Social protection and risk management'}, {'code': '6', 'name': ''}]
6 [{'code': '2', 'name': 'Public sector governance'}, {'code': '4', 'name': 'Financial and private sector development'}]
7 [{'code': '11', 'name': 'Environment and natural resources management'}, {'code': '8', 'name': ''}]
8 [{'code': '10', 'name': 'R

In [49]:
#Normalize to cretae table and read the data
data_normal = json_normalize(wb_json_data, 'mjtheme_namecode')
#read data in a dataframe
dfdirup = pd.DataFrame(data_normal)
dfdirup.head(10)

code                                          name
0    8                             Human development
1   11                                              
2    1                           Economic management
3    6         Social protection and risk management
4    5                         Trade and integration
5    2                      Public sector governance
6   11  Environment and natural resources management
7    6         Social protection and risk management
8    7                   Social dev/gender/inclusion
9    7                   Social dev/gender/inclusion

In [50]:
#Find the largest values from the sorted group
totals = dfdirup.name.value_counts()
print("The 10 largest project themes:")
totals.head(10)

The 10 largest project themes:


Environment and natural resources management    223
Rural development                               202
Human development                               197
Public sector governance                        184
Social protection and risk management           158
Financial and private sector development        130
                                                122
Social dev/gender/inclusion                     119
Trade and integration                            72
Urban development                                47
Name: name, dtype: int64

Above results are displaying the count of unique values of project themes and most are "Environment and natural resources management", interesting we have missing value with 122 count.

# In 2. above you will notice that some entries have only the code and the name is missing. Create a dataframe with the missing names filled in.

In [51]:
# update missing names in dataframe by the code and value
dfdirup.groupby(['code', 'name']).size()
dfdirup.loc[dfdirup.code == '1', 'name'] = 'Economic Management'
dfdirup.loc[dfdirup.code == '2', 'name'] = 'Public sector governance'
dfdirup.loc[dfdirup.code == '3', 'name'] = 'Rule of law'
dfdirup.loc[dfdirup.code == '4', 'name'] = 'Financial and private sector development'
dfdirup.loc[dfdirup.code == '5', 'name'] = 'Trade and integration'
dfdirup.loc[dfdirup.code == '6', 'name'] = 'Social protection and risk management'
dfdirup.loc[dfdirup.code == '7', 'name'] = 'Social dev/gender/inclusion'
dfdirup.loc[dfdirup.code == '8', 'name'] = 'Human development'
dfdirup.loc[dfdirup.code == '9', 'name'] = 'Urban development'
dfdirup.loc[dfdirup.code == '10', 'name'] = 'Rural development'
dfdirup.loc[dfdirup.code == '11', 'name'] = 'Environment and natural resources management'

In [52]:
print('Top 10 major project themes with the missing names filled in:')
print(dfdirup.name.value_counts().head(10))

Top 10 major project themes with the missing names filled in:
Environment and natural resources management    250
Rural development                               216
Human development                               210
Public sector governance                        199
Social protection and risk management           168
Financial and private sector development        146
Social dev/gender/inclusion                     130
Trade and integration                            77
Urban development                                50
Economic Management                              38
Name: name, dtype: int64


In the above results, the missing values are been updated and the order is almost same which we have seen in above results with missing data, the "Environment and natural resources management" has been updated and leading the group. Also, we notice that Economic Management has been apprearing in the list after the updates amount top 10.

## Its is another Alternate way for filling the missing values.

In [39]:
#with open('data/world_bank_projects.json') as json_file:
#    json_data = json.load(json_file)
#normalize to flatten the data
data_normal = json_normalize(wb_json_data, 'mjtheme_namecode')
#read data into a dataframe
dfnjname = pd.DataFrame(data_normal)
dfnjname.head(10)

code                                          name
0    8                             Human development
1   11                                              
2    1                           Economic management
3    6         Social protection and risk management
4    5                         Trade and integration
5    2                      Public sector governance
6   11  Environment and natural resources management
7    6         Social protection and risk management
8    7                   Social dev/gender/inclusion
9    7                   Social dev/gender/inclusion

In [40]:
#sort into groups by name so we can find the largest
totals = dfnjname.name.value_counts()
print("The 10 largest project themes:")
totals.head(10)

The 10 largest project themes:


Environment and natural resources management    223
Rural development                               202
Human development                               197
Public sector governance                        184
Social protection and risk management           158
Financial and private sector development        130
                                                122
Social dev/gender/inclusion                     119
Trade and integration                            72
Urban development                                47
Name: name, dtype: int64

In [41]:
#map out the theme names to codes
name_code_dict = {}
for row in dfnjname.itertuples():
    if row[2] != '':
        name_code_dict[row[1]] = row[2]
        
name_code_dict

{'8': 'Human development',
 '1': 'Economic management',
 '6': 'Social protection and risk management',
 '5': 'Trade and integration',
 '2': 'Public sector governance',
 '11': 'Environment and natural resources management',
 '7': 'Social dev/gender/inclusion',
 '4': 'Financial and private sector development',
 '10': 'Rural development',
 '9': 'Urban development',
 '3': 'Rule of law'}

In [42]:
# Fill in missing theme names using the name dictionary
for row in dfnjname.itertuples():
    if row[2] == '':
        dfnjname.set_value(row[0], 'name', name_code_dict[row[1]])

/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:4: FutureWarning: set_value is deprecated and will be removed in a future release. Please use .at[] or .iat[] accessors instead
  after removing the cwd from sys.path.


In [44]:
print('Top 10 major project themes with the missing names filled in:')
print(dfnjname.name.value_counts().head(10))

Top 10 major project themes with the missing names filled in:
Environment and natural resources management    250
Rural development                               216
Human development                               210
Public sector governance                        199
Social protection and risk management           168
Financial and private sector development        146
Social dev/gender/inclusion                     130
Trade and integration                            77
Urban development                                50
Economic management                              38
Name: name, dtype: int64
